### PR TITLE
Fix use of time_t internally and APIs

### PR DESCRIPTION
**Pull request by tmshort**
This changes internal (libssl) times to use ossl_time.

Replacement public APIs (for backwards compatibility) are introduced to
support time_t in the interface (rather than long).

##### Checklist

- [X] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Seems like some overlap between this and #19082...

**Comment by tmshort** [comment]
> Seems like some overlap between this and #19082...

Yes, I must've missed that one!

File: apps/s_time.c [hunk 1]
```diff
@@ -32,6 +33,12 @@
 #define SECONDS 30
 #define SECONDSSTR "30"
 
+/* not as accurate as in the library, but no worse than original code */
+OSSL_TIME ossl_time_now(void)
```
Comment (paulidale) [inline apps/s_time.c:37]: Why not include crypto/time.c and get the accurate version?

File: doc/man3/OPENSSL_gmtime.pod [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+with B<timer> equal to NULL means use the current time.
+
+OPENSSL_gmtime_adj() converts B<tm> into a days and seconds value, adds the
+offsets, then converts back into a B<struct tm> specified by B<tm>. Leap seconds
+are not considered.
+
+OPENSSL_gmtime_diff() calculates the difference between the two B<struct tm>
+structures B<from> and B<to>. The difference in days is placed into B<pday>,
+the remaining seconds are placed to B<psec>. The value in B<psec> will be less
+than the number of seconds per day (3600). Leap seconds are not considered.
```
Comment (paulidale) [inline doc/man3/OPENSSL_gmtime.pod:40]: 86400 seconds in a day not 3600.

File: doc/man3/SSL_CTX_set_timeout.pod [hunk 3]
```diff
@@ -2,21 +2,27 @@
...
+SSL_CTX_set_timeout_ex,
+SSL_CTX_get_timeout_ex - manipulate timeout values for session caching
 
 =head1 SYNOPSIS
 
  #include <openssl/ssl.h>
 
  long SSL_CTX_set_timeout(SSL_CTX *ctx, long t);
  long SSL_CTX_get_timeout(SSL_CTX *ctx);
+ time_t SSL_CTX_set_timeout_ex(SSL_CTX *ctx, time_t t);
```
Comment (paulidale) [inline doc/man3/SSL_CTX_set_timeout.pod:15]: The timeout is a duration isn't it.  As such, I don't see a need to accept a `time_t`.  `long` is more than sufficient.

File: doc/man3/SSL_SESSION_get_time.pod [hunk 4]
```diff
@@ -16,6 +20,11 @@ SSL_get_time, SSL_set_time, SSL_get_timeout, SSL_set_timeout
  long SSL_SESSION_get_timeout(const SSL_SESSION *s);
  long SSL_SESSION_set_timeout(SSL_SESSION *s, long tm);
 
+ time_t SSL_SESSION_get_time_ex(const SSL_SESSION *s);
+ time_t SSL_SESSION_set_time_ex(SSL_SESSION *s, time_t tm);
+ time_t SSL_SESSION_get_timeout_ex(const SSL_SESSION *s);
```
Comment (paulidale) [inline doc/man3/SSL_SESSION_get_time.pod:25]: Again, not sure we need or want timeouts using `time_t`.

File: doc/man3/SSL_get_default_timeout.pod [hunk 5]
```diff
@@ -2,30 +2,36 @@
...
-SSL_get_default_timeout - get default session timeout value
+SSL_get_default_timeout,
+SSL_get_default_timeout_ex - get default session timeout value
 
 =head1 SYNOPSIS
 
  #include <openssl/ssl.h>
 
  long SSL_get_default_timeout(const SSL *ssl);
+ time_t SSL_get_default_timeout_ex(const SSL *ssl);
```
Comment (paulidale) [inline doc/man3/SSL_get_default_timeout.pod:13]: Same here.

**Review by paulidale — COMMENTED** [review]
Looks good and there are some changes here that aren't in #19082

File: doc/man3/OPENSSL_gmtime.pod (hunk 2 above)
Comment (tmshort) [reply doc/man3/OPENSSL_gmtime.pod:40]: Nah... I'm just going to restrict it to an hour... whoops.

File: doc/man3/SSL_CTX_set_timeout.pod (hunk 3 above)
Comment (tmshort) [reply doc/man3/SSL_CTX_set_timeout.pod:15]: Duration and let call it "calandar time" are related but distinct concepts. Duration is not fixed in time, whereas a calendar time is. However, calendar time is nothing but a duration from a fixed point in time. So, they really could be the same type. Both are effectively bound by the same constraints, one of which is the size/type of storage. The OSSL_TIME code itself allows for subtraction (which results in a duration) and addition (one of which must be a duration). `sleep(3)` takes a duration in seconds, as an `unsigned int`. But I swear there was a type specific for durations, but I can't find it.

**Comment by kroeckx** [comment]
Adding new functions with a time_t seems wrong. We should deprecate all public function with a time_t. If time_t is only used internally, we can fix Y2038 problems.

**Comment by rsbeckerca** [comment]
We should not use time_t internally as there are 32-bit builds that will not work in Epoch 1.

**Comment by kroeckx** [comment]
To clarify, we should only use time_t when needed to talk to the OS.
Some OSs provide multiple interfaces, and might have one that doesn't
have the Y2038 problem. That might be multiple interfaces that use
time_t. Some OSs will not provide a Y2038 compatible interface, and I'm
not sure if that's up to us to fix or not.

File: doc/man3/SSL_CTX_set_timeout.pod (hunk 3 above)
Comment (paulidale) [reply doc/man3/SSL_CTX_set_timeout.pod:15]: There are more than a few variations documented:

| Call | Type |
| --- | ---: |
| `alarm(2)` | `unsigned int` |
| `clock(2)` | `clock_t` |
| `clock_gettime(2)` | `struct timespec` |
| `epoll_wait(2)` | `int` |
| getitimer(2)` | `struct timeval` |
| `poll(2)` | `int` |
| `sleep(3)` | `unsigned int` |
| `usleep(3)` | `useconds_t` |

I can't locate a specific duration type either.

File: doc/man3/SSL_SESSION_get_time.pod [hunk 6]
```diff
@@ -39,6 +48,10 @@ to B<tm>.
 The SSL_get_time(), SSL_set_time(), SSL_get_timeout(), and SSL_set_timeout()
 functions are synonyms for the SSL_SESSION_*() counterparts.
 
+SSL_SESSION_get_time_ex(), SSL_SESSION_set_time_ex(),
+SSL_SESSION_get_timeout_ex(), and SSL_SESSION_set_timeout_ex() functions are
+synonyms for the SSL_SESSION_*() (without the "_ex") counterparts.
```
Comment (kroeckx) [inline doc/man3/SSL_SESSION_get_time.pod:53]: I understand that as: #define SSL_SESSION_get_time_ex SSL_SESSION_get_time

**Comment by kroeckx** [comment]
It seems to me that most of the functions here actually need a time_t:
- It's about a timeout, which is relative, not some specific time.
- SSL_CTX_flush_sessions also seems to be used relative to the current
  time, and maybe it should just take a relative time instead of a
  time_t.

The only exception is SSL_SESSION_get_time and related functions, which
take a specific point in time. This can be changed to a type that is
Y2038 compatible, it does not need to be a time_t.

**Comment by tmshort** [comment]
I'm abandoning this, but going to create a PR for the `OPENSSL_gmtime` functions.

**Pull request closed without merging**